File: kotlin-multiplatform-app/composeApp/src/commonMain/kotlin/pl/edu/pb/kotlinmultiplatformapp/data/database/repository/DatabaseRepositoryImpl.kt
package pl.edu.pb.kotlinmultiplatformapp.data.database.repository

import pl.edu.pb.kotlinmultiplatformapp.data.database.local.DatabaseDataSource
import pl.edu.pb.kotlinmultiplatformapp.data.infinitelist.remote.InfiniteListDataSource
import pl.edu.pb.kotlinmultiplatformapp.data.model.CharacterEntity
import pl.edu.pb.kotlinmultiplatformapp.model.Character

class DatabaseRepositoryImpl(
    private val infiniteListDataSource: InfiniteListDataSource,
    private val databaseDataSource: DatabaseDataSource,
) : DatabaseRepository {
    override suspend fun getAllCharacters(): List<Character> {
        return databaseDataSource.getAllCharacters().map {
            Character(
                id = it.id.toString(),
                name = it.name,
                status = it.status,
                species = it.species,
                gender = it.gender,
                image = it.image,
            )
        }
    }

    override suspend fun saveAllCharacters() {
        var url: String? = "https://rickandmortyapi.com/api/character"
        val allCharacters = mutableListOf<CharacterEntity>()
        do {
            val charactersPage = infiniteListDataSource.getCharactersPage(url!!)
            allCharacters += charactersPage.results
            url = charactersPage.info.next
        } while (url != null)
        databaseDataSource.saveAllCharacters(allCharacters)
    }

    override suspend fun clearDatabase() {
        databaseDataSource.clearDatabase()
    }
}

File: flutter-multiplatform-app/flutter_multiplatform_app/lib/data/database/repository/database_repository_impl.dart
import 'package:flutter_multiplatform_app/data/database/local/database_data_source.dart';
import 'package:flutter_multiplatform_app/data/database/repository/database_repository.dart';
import 'package:flutter_multiplatform_app/data/infinitylist/remote/infinite_list_data_source.dart';
import 'package:flutter_multiplatform_app/data/model/character_entity.dart';
import 'package:flutter_multiplatform_app/model/character.dart';

class DatabaseRepositoryImpl implements DatabaseRepository {
  DatabaseRepositoryImpl({required DatabaseDataSource databaseDataSource, required InfiniteListDataSource infiniteListDataSource})
      : _databaseDataSource = databaseDataSource,
        _infiniteListDataSource = infiniteListDataSource;

  final DatabaseDataSource _databaseDataSource;
  final InfiniteListDataSource _infiniteListDataSource;

  @override
  Future<List<Character>> getAllCharacters() async {
    final characterEntities = await _databaseDataSource.getAllCharacters();

    return characterEntities
        .map((entity) => Character(
            id: entity.id,
            name: entity.name,
            status: entity.status,
            species: entity.species,
            gender: entity.gender,
            image: entity.image))
        .toList();
  }

  @override
  Future<void> saveAllCharacters() async {
    String? url = "https://rickandmortyapi.com/api/character";
    late List<CharacterEntity> allCharacters = [];
    do {
      final charactersPage = await _infiniteListDataSource.getCharactersPage(url!);
      allCharacters += charactersPage.results;
      url = charactersPage.info.next;
    } while (url != null);
    _databaseDataSource.saveAllCharacters(allCharacters);
  }

  @override
  Future<void> clearDatabase() async {
    _databaseDataSource.clearDatabase();
  }
}
